lib/job/arUpdatePublicationStatusJob.class.php

```diff
@@ -18,39 +18,129 @@
  */
 
 /**
- * Updates the publication status to the descendants of an information object.
+ * Updates the publication status to the descendants of an information object. Uses
+ * ES updateByQuery to update the ES index rather than replace each child document.
  */
 class arUpdatePublicationStatusJob extends arBaseJob
 {
     /**
      * @see arBaseJob::$requiredParameters
      */
-    protected $extraRequiredParameters = ['objectId', 'publicationStatusId'];
+    protected $extraRequiredParameters = ['objectId', 'publicationStatusId', 'updateDescendants'];
 
     public function runJob($parameters)
     {
+        $validPublicationStatusList = [
+            QubitTerm::PUBLICATION_STATUS_DRAFT_ID,
+            QubitTerm::PUBLICATION_STATUS_PUBLISHED_ID,
+        ];
+
+        if (null === $publicationStatus = QubitTerm::getById($parameters['publicationStatusId'])) {
+            $this->error($this->i18n->__('Invalid publication status id: %1', ['%1' => $parameters['publicationStatusId']]));
+
+            return false;
+        }
+
         if (null === $resource = QubitInformationObject::getById($parameters['objectId'])) {
             $this->error($this->i18n->__('Invalid description id: %1', ['%1' => $parameters['objectId']]));
 
             return false;
         }
 
-        if (null === $publicationStatus = QubitTerm::getById($parameters['publicationStatusId'])) {
-            $this->error($this->i18n->__('Invalid publication status id: %1', ['%1' => $parameters['publicationStatusId']]));
+        $sql = 'UPDATE status
+            SET status.status_id = :publicationStatus
+            WHERE status.type_id = :publicationStatusTypes ';
 
-            return false;
+        if ($parameters['updateDescendants']) {
+            $sql .= 'AND status.object_id IN (
+                SELECT des.id
+                FROM information_object io
+                JOIN information_object des
+                    ON des.lft >= io.lft AND des.lft < io.rgt
+                JOIN slug
+                    ON io.id = slug.object_id
+                WHERE slug.slug = :slug
+            )';
+
+            $message = $this->i18n->__(
+                'Updating publication status for "%1" and descendants to "%2".',
+                ['%1' => $resource->getTitle(['cultureFallback' => true]), '%2' => $publicationStatus->name]
+            );
+        } else {
+            $sql .= 'AND status.object_id IN (
+                SELECT io.id
+                FROM information_object io
+                JOIN slug
+                    ON io.id = slug.object_id
+                WHERE slug.slug = :slug
+            )';
+
+            $message = $this->i18n->__(
+                'Updating publication status for "%1" to "%2".',
+                ['%1' => $resource->getTitle(['cultureFallback' => true]), '%2' => $publicationStatus->name]
+            );
```

Comment:
I see you took the query from the gist, which was created to facilitate the work in support (using the slug) and without a lot of care.

If you keep the resource update outside the job and only trigger it for the descendants update this could be a lot more straight forward. And I wouldn't use a subquery, something like ...

```sql
UPDATE status
JOIN information_object io ON status.object_id = io.id
SET status.status_id = :publicationStatus
WHERE status.type_id = :publicationStatusType
AND io.lft > :lft AND io.rgt < :rgt;
```

If you want to keep the resource update here you could use `AND io.lft >= :lft AND io.rgt <= :rgt;`.

Reply:
I have refactored this in latest commit.